### ISSUE TITLE
Edge Ngram Token filter should also include tokens that are shorter than the min_gram

### ISSUE THREAD

**Issue by dougnelas** [elastic/elasticsearch#36012]
**Describe the feature**:  

NEdgeGram token filter should also emit tokens that are shorter than the min_gram setting.   The current behavior is not to include tokens that are shorter than the min_gram setting.  For example when the min_gram is set to three, tokens that are less than three characters are not emitted.  This is forcing users to use very small settings to make shorter tokens searchable.

There is also a `preserveOriginal` setting but setting this to true does not change the output.  However, while this would solve the shorter token problem, it would also include tokens that are greater than the max_gram size which may not be desirable.

**Comment by jimczi** [comment]
`preserveOriginal` is not exposed in ES yet so it cannot change the output. Can you explain why it would be an issue to also add tokens greater than the max_gram size ?
Also note that the `text` field has a new option called `index_prefixes` which can be used to index prefixes. Prefix queries on a field that has this option activated will automatically switch to the prefix field if the number of characters in the prefix is in the range of the `index_prefixes` field.

**Comment by dougnelas** [comment]
Edgengrams are useful for typeahead.  In order to reduce the number of
tokens it is nice to be able to set the min gram size to match the number
of characters entered before the search.  Also in most typeahead cases I
would set this to min 3. max 5 or 6. This is very efficient as it really
limits the number of chars that need to be matched.  With edge grams you
need a length of at least 5 to work properly.  If I set the min gram to 1
now my set is significantly larger.

**Comment by polyfractal** [comment]
I re-read the original request, and I think I misunderstood the first time I read through it.

@dougnelas: are you wanting the filter to emit tokens if they are less than the `min_gram` size if they are below the `min_gram` size _before_ tokenization happens?

E.g. if `min_gram: 3`, the token `"ok"` would be dropped because it is below the threshold.  And `preserveOriginal` won't work, because that will start to emit long tokens too (`"foobar"`).  So to get the smaller tokens, you have to set `min_gram` lower which has the bad side effect of also gram'ing all the longer tokens, bloating index and making the grams less specific.

Basically, the behavior you're requesting is:

- If a token is > `min_gram`, chop it up with `min_gram`/`max_gram` settings
- If a token is < `min_gram`, emit the token instead of dropping it.

Is that correct?

**Comment by romseygeek** [comment]
@dougnelas does the conditional token filter solve your particular use-case?

**Comment by astefan** [comment]
@dougnelas, there is no further feedback from your side for quite some time now. Since we do have a suggested solution for you, I will go ahead and close this issue.

**Comment by suikast42** [comment]
That solves my problem ;-) 

>  "filter": [ "edge_4_255_filter" ],
      "script": {
        "source": "token.getTerm().length() >= 4"
      }
          },
          "edge_4_255_filter": {
            "type": "edge_ngram",
            "min_gram": 4,
            "max_gram": 255
          }
        },

**Comment by pkgonan** [comment]
@suikast42 Hi !
Do you have fully working code about that ?

I tried but it is not works. 
Is there anything wrong with me?

```
	"settings": {
		"index.max_ngram_diff": 3,
	  	"analysis": {
	    		"analyzer": {
	      			"fulltext": {
	        			"tokenizer": "whitespace",
					"filter": ["trim", "ngram_filter"],
					"script": {
						"source": "token.getTerm().length() < 100"
					}
	      			}
	    		},
        		"filter": {
          			"ngram_filter": {
            				"type": "ngram",
            				"min_gram": 2,
            				"max_gram": 5
          			}
        		}
	  	}
  	},
	"mappings": {
		"dynamic_templates": [
			{
				"strings": {
					"match_mapping_type": "string",
					"mapping": {
						"type": "text",
						"analyzer": "fulltext",
						"fields": {
              						"raw": {
                						"type":  "keyword",
                						"ignore_above": 256
              						}
            					}
					}
				}
			}
		]
	}
```
